pkg/controller/certrotation/certrotation_controller.go

```diff
@@ -0,0 +1,463 @@
+		go certRotator.Run(ctx, workers)
+	}
+
+	<-ctx.Done()
+}
+
+func getServerIPsFromInfra(cfg *configv1.Infrastructure) []string {
+	if cfg.Status.PlatformStatus == nil {
+		return []string{}
+	}
+	switch cfg.Status.PlatformStatus.Type {
+	case configv1.BareMetalPlatformType:
+		if cfg.Status.PlatformStatus.BareMetal == nil {
+			return []string{}
+		}
+		return cfg.Status.PlatformStatus.BareMetal.APIServerInternalIPs
+	case configv1.OvirtPlatformType:
+		if cfg.Status.PlatformStatus.Ovirt == nil {
+			return []string{}
+		}
+		return cfg.Status.PlatformStatus.Ovirt.APIServerInternalIPs
+	case configv1.OpenStackPlatformType:
+		if cfg.Status.PlatformStatus.OpenStack == nil {
+			return []string{}
+		}
+
+		return cfg.Status.PlatformStatus.OpenStack.APIServerInternalIPs
+	case configv1.VSpherePlatformType:
+		if cfg.Status.PlatformStatus.VSphere == nil {
+			return []string{}
+		}
+		if cfg.Status.PlatformStatus.VSphere.APIServerInternalIPs != nil {
+			return cfg.Status.PlatformStatus.VSphere.APIServerInternalIPs
+		}
+		return cfg.Status.PlatformStatus.VSphere.APIServerInternalIPs
+	case configv1.NutanixPlatformType:
+		if cfg.Status.PlatformStatus.Nutanix == nil {
+			return []string{}
+		}
+		return cfg.Status.PlatformStatus.Nutanix.APIServerInternalIPs
+	default:
+		return []string{}
+	}
+}
+
+func (c *CertRotationController) PreFlightChecks() error {
+
+	// Ensure machine-config-server-tls secret is of the kubernetes.io/tls type. The cert-rotation-controller expects this type
+	// On a cluster where the MCS CA/cert is not managed, the secret will be of the opaque type, and we'll have to do this one
+	// time change.
+
+	// Retrieve the machine-config-server-tls secret
+	mcsTLSSecret, err := c.kubeClient.CoreV1().Secrets(ctrlcommon.MCONamespace).Get(context.Background(), ctrlcommon.MachineConfigServerTLSSecretName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot get MCS TLS secret: %w", err)
+	}
+
+	// Check if the secret type needs to be updated to kubernetes.io/tls
+	if mcsTLSSecret.Type != corev1.SecretTypeTLS {
+		klog.Infof("Migration to %s for %s required", corev1.SecretTypeTLS, ctrlcommon.MachineConfigServerTLSSecretName)
+
+		// Delete the existing secret
+		if err := c.kubeClient.CoreV1().Secrets(ctrlcommon.MCONamespace).Delete(context.Background(), ctrlcommon.MachineConfigServerTLSSecretName, metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+		// Clone a new secret of the correct type, with the same data
+		newSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ctrlcommon.MachineConfigServerTLSSecretName,
+				Namespace: ctrlcommon.MCONamespace,
+			},
+			Data: mcsTLSSecret.Data,
+			Type: corev1.SecretTypeTLS,
+		}
+		// Recreate the secret. is this necessary? It will be updated in the first loop
+		if _, err := c.kubeClient.CoreV1().Secrets(ctrlcommon.MCONamespace).Create(context.Background(), newSecret, metav1.CreateOptions{}); err != nil {
+			return err
```

Comment:
Unfortunately, type is immutable, I tried doing a patch first :/ 

Worst case if this create fails, the cert controller will immediately create a new one, so the cluster should not be missing this secret for longer than the sync window of this controller(1 minute or less). If the delete fails, we may be in trouble though, as the old secret will "stop" our controller from coming up. 

Additionally, since this secret is not in the format(no annotations) expected by the cert controller, I've noticed it gets updated immediately by the controller - hence my comment above asking "is this even required?". I leaned to the safe side trying to "convert" the old secret, but it may be pointless.